snuba/pipeline/single_query_plan_pipeline.py

```diff
@@ -1,58 +1,97 @@
-    A query pipeline for a single query plan.
-
-    TODO: Currently only query plan building and processing is done by the query
-    pipeline, eventually the rest of the query processing sequence will move into
-    the pipeline as well.
+    Executes the processing phase of a single plan query. Which means a
+    query based on a single entity, that would produce a query plan based
+    on a single storage.
     """
 
     def __init__(
         self,
-        request: Request,
-        runner: QueryRunner,
         query_plan_builder: ClickhouseQueryPlanBuilder,
-    ):
-        self.__request = request
-        self.__runner = runner
+        processors: ClickhouseProcessorsExecutor,
+    ) -> None:
         self.__query_plan_builder = query_plan_builder
+        self.__entity_processors = EntityProcessorsExecutor()
+        self.__clickhouse_processors = processors
+
+    def execute(
+        self, query: LogicalQuery, settings: RequestSettings
+    ) -> ClickhouseQueryPlan:
+        self.__entity_processors.execute(query, settings)
+
+        query_plan = self.__query_plan_builder.build_plan(query, settings)
+        self.__clickhouse_processors.execute(query_plan, settings)
+        return query_plan
 
-    def execute(self) -> QueryResult:
-        # Execute entity processors
-        _execute_entity_processors(self.__request)
 
-        # Build and execute query plan
-        self.__query_plan = self.__query_plan_builder.build_plan(self.__request)
-        _execute_query_plan_processors(self.__query_plan, self.__request)
-        return self.__query_plan.execution_strategy.execute(
-            self.__query_plan.query, self.__request.settings, self.__runner
+class PreStrategyProcessingPipeline(SinglePlanProcessingPipeline):
+    """
+    Executes the query processing of a single query plan query up to the
+    plan query processors. It does not execute the db query processors
+    since they are ran by the execution strategy.
+
+    This is the processing pipeline to use when executing a simple query
+    that references a single entity and a single storage.
+    """
+
+    def __init__(self, query_plan_builder: ClickhouseQueryPlanBuilder) -> None:
+        super().__init__(
+            query_plan_builder,
+            ClickhouseProcessorsExecutor(lambda plan: plan.plan_processors),
         )
 
-    @property
-    def query_plan(self) -> ClickhouseQueryPlan:
-        return self.__query_plan
+
+class SinglePlanExecutionPipeline(QueryExecutionPipeline):
+    """
+    Executes a simple (single entity) query.
+    This relies on a PreStrategyProcessingPipeline to perform the processing
+    and then relies on the execution strategy provided by the query plan to
+    actually execute the query.
+    """
+
+    def __init__(
+        self, runner: QueryRunner, query_plan_builder: ClickhouseQueryPlanBuilder,
+    ):
+        self.__runner = runner
+        self.__processing_pipeline = PreStrategyProcessingPipeline(query_plan_builder)
```

Comment:
It seems like an anti pattern to me to have a stage of the pipeline execute another pipeline. Having different pipelines that are all thin wrappers around other pipelines already confuses me without also having those pipelines take other pipelines as arguments and execute them.

Reply:
> It seems like an anti pattern to me to have a stage of the pipeline execute another pipeline.

I don't think I would agree with this. Why would it be an anti pattern? Pipelines can be composed and each component can have its own structure.
The issue is that when we throw composite queries and the delegator in the mix, our execution pipeline will have to do a lot of steps and can take different paths, many of them are similar. I think making them modular is inevitable. 

Also I don't think this is a thin wrapper around the processing pipeline. The execution pipeline adds a major step to the processing one, which is the execution of the query. The processing pipeline is its own module since it has to be used independently for the composite execution pipeline (which does not exist in this PR yet).

Would it help if I added trhe composite pipeline in this one as well (even though incomplete) ?

Reply:
Now this is quite different than it was, so probably the specific comment does not apply. Though we will need to compose pipelines from smaller parts anyway (the composite processing pipeline needs to run several entity processing pipelines (one per entity involved).